dsgrid/spark/functions.py

```diff
@@ -0,0 +1,532 @@
+
+
+def create_temp_view(df: DataFrame) -> str:
+    """Create a temporary view with a random name and return the name."""
+    view1 = make_temp_view_name()
+    df.createOrReplaceTempView(view1)
+    return view1
+
+
+def make_temp_view_name() -> str:
+    """Make a random name to be used as a view."""
+    return f"{TEMP_TABLE_PREFIX}_{uuid4().hex}"
+
+
+def drop_temp_tables_and_views() -> None:
+    """Drop all temporary views and tables."""
+    drop_temp_views()
+    drop_temp_tables()
+
+
+def drop_temp_tables() -> None:
+    """Drop all temporary tables."""
+    if not use_duckdb():
+        return
+
+    spark = get_spark_session()
+    query = f"SELECT * FROM pg_tables WHERE tablename LIKE '{TEMP_TABLE_PREFIX}%'"
+    for row in spark.sql(query).collect():
+        spark.sql(f"DROP TABLE {row.tablename}")
+        logger.debug("Dropped temp table %s", row.tablename)
+
+
+def drop_temp_views() -> None:
+    """Drop all temporary views."""
+    if not use_duckdb():
+        return
+
+    spark = get_spark_session()
+    query = "SELECT view_name FROM duckdb_views() WHERE NOT internal AND view_name LIKE '{TEMP_TABLE_PREFIX}%'"
+    for row in spark.sql(query).collect():
+        spark.sql(f"DROP VIEW {row.view_name}")
+        logger.debug("Dropped temp view %s", row.view_name)
+
+
+def cross_join(df1: DataFrame, df2: DataFrame) -> DataFrame:
+    """Return a cross join of the two dataframes."""
+    if use_duckdb():
+        view1 = create_temp_view(df1)
+        view2 = create_temp_view(df2)
+        spark = get_spark_session()
+        return spark.sql(f"SELECT {view1}.*, {view2}.* FROM {view1} CROSS JOIN {view2}")
+
+    return df1.crossJoin(df2)
+
+
+def except_all(df1: DataFrame, df2: DataFrame) -> DataFrame:
+    """Return a dataframe with all rows in df1 that are not in df2."""
+    method = _except_all_duckdb if use_duckdb() else _except_all_spark
+    return method(df1, df2)
+
+
+def _except_all_duckdb(df1: DataFrame, df2: DataFrame) -> DataFrame:
+    view1 = create_temp_view(df1)
+    view2 = create_temp_view(df2)
+    query = f"""
+        SELECT * FROM {view1}
+        EXCEPT ALL
+        SELECT * FROM {view2}
+    """
+    spark = get_spark_session()
+    return spark.sql(query)
+
+
+def _except_all_spark(df1: DataFrame, df2: DataFrame) -> DataFrame:
+    return df1.exceptAll(df2)
+
+
+def shift_time_zone(
+    df: DataFrame, time_column: str, from_time_zone: str, to_time_zone: str, new_column: str
+) -> DataFrame:
+    """Shift the time zone of the time_column. new_column can be the same as time_column."""
+    if use_duckdb():
+        view = create_temp_view(df)
+        cols = df.columns[:]
+        if time_column == new_column:
+            cols.remove(time_column)
+        cols_str = ",".join(cols)
+        query = f"""
+            SELECT
+                {cols_str},
+                CAST(timezone('{to_time_zone}', {time_column}) AS TIMESTAMPTZ) AS {new_column}
```

Comment:
Need to change the Spark side to use `convert_time_zone`

Reply:
Okay, so because the time column is TIMESTAMPTZ type. It needs to be this:

```python
query = f"""
             SELECT
                 {cols_str},
                 CAST(timezone('{to_time_zone}',
                     CAST(timezone('{from_time_zone}', {time_column}) AS TIMESTAMP)
                            ) AS TIMESTAMPTZ) AS {new_column}
              FROM {view}"""
```